Detectors/EMCAL/reconstruction/src/CaloRawFitterGamma2.cxx

```diff
@@ -0,0 +1,172 @@
+// Copyright CERN and copyright holders of ALICE O2. This software is
+// distributed under the terms of the GNU General Public License v3 (GPL
+// Version 3), copied verbatim in the file "COPYING".
+//
+// See http://alice-o2.web.cern.ch/license for full licensing information.
+//
+// In applying this license CERN does not waive the privileges and immunities
+// granted to it by virtue of its status as an Intergovernmental Organization
+// or submit itself to any jurisdiction.
+
+/// \file CaloRawFitterGamma2.cxx
+/// \author Martin Poghosyan (Martin.Poghosyan@cern.ch)
+
+#include "FairLogger.h"
+#include <random>
+
+// ROOT sytem
+#include "TMath.h"
+
+#include "EMCALReconstruction/Bunch.h"
+#include "EMCALReconstruction/CaloFitResults.h"
+#include "DataFormatsEMCAL/Constants.h"
+
+#include "EMCALReconstruction/CaloRawFitterGamma2.h"
+
+using namespace o2::emcal;
+
+CaloRawFitterGamma2::CaloRawFitterGamma2() : CaloRawFitter("Chi Square ( Gamma2 )", "Gamma2")
+{
+  mAlgo = FitAlgorithm::Gamma2;
+}
+
+CaloFitResults CaloRawFitterGamma2::evaluate(const std::vector<Bunch>& bunchlist,
+                                             std::optional<unsigned int> altrocfg1, std::optional<unsigned int> altrocfg2)
+{
+  float time = 0;
+  float amp = 0;
+  float chi2 = 0;
+  int ndf = 0;
+  bool fitDone = false;
+
+  auto [nsamples, bunchIndex, ampEstimate,
+        maxADC, timeEstimate, pedEstimate, first, last] = preFitEvaluateSamples(bunchlist, altrocfg1, altrocfg2, mAmpCut);
+
+  if (ampEstimate >= mAmpCut) {
+    time = timeEstimate;
+    int timebinOffset = bunchlist.at(bunchIndex).getStartTime() - (bunchlist.at(bunchIndex).getBunchLength() - 1);
+    amp = ampEstimate;
+
+    if (nsamples > 2 && maxADC < constants::OVERFLOWCUT) {
+      doParabolaFit(timeEstimate - 1, amp, time);
+      mNiter = 0;
+      fitDone = doFit_1peak(first, nsamples, amp, time, chi2);
+
+      if (!fitDone) {
+        amp = ampEstimate;
+        time = timeEstimate;
+        chi2 = 1.e9;
+      }
+
+      time += timebinOffset;
+      timeEstimate += timebinOffset;
+      ndf = nsamples - 2;
+    }
+  }
+
+  if (fitDone) {
+    float ampAsymm = (amp - ampEstimate) / (amp + ampEstimate);
+    float timeDiff = time - timeEstimate;
+
+    if ((TMath::Abs(ampAsymm) > 0.1) || (TMath::Abs(timeDiff) > 2)) {
+      amp = ampEstimate;
+      time = timeEstimate;
+      fitDone = false;
+    }
+  }
+  if (amp >= mAmpCut) {
+    if (!fitDone) {
+      std::default_random_engine generator;
+      std::uniform_real_distribution<float> distribution(0.0, 1.0);
+      amp += (0.5 - distribution(generator));
+    }
+    time = time * constants::EMCAL_TIMESAMPLE;
+    time -= mL1Phase;
+
+    return CaloFitResults(maxADC, pedEstimate, mAlgo, amp, time, (int)time, chi2, ndf);
+  }
+  return CaloFitResults(-1, -1);
+}
+
+bool CaloRawFitterGamma2::doFit_1peak(int ifirst, int nS, float& A0, float& t0, float& chi2)
```

Comment:
Variable names should be meaningful. Function parameters should be input, tuple of the fit result should be output. In case of non-convergence the function should throw an exception (which must be handled) rather than a boolean return value.

Reply:
About the exceptions, is it really needed. Because one exception will crash the whole process. I think it should be boolean so it ignores the event and go to the next. What do you think ??

Reply:
Yes, it is needed. The division-by-0 case will leave all parameters which are function arguments in this implementation in undefined state. Users must be forced to catch it. Having a non-crashing code should not be the focus if the price is uninitialized values. Using int / bool for error handling is not an option, at least not in EMCAL code.